Update and enhance the test environment

gildegoma opened the pull request:
As a prerequisite to further works (e.g. rebasing #52), I propose here to refresh the test environment with following changes:

- Replace outdated versions of CentOS, Ubuntu and Debian by new versions
- Modify `tests/assertions.yml` to provide stronger validation of **rvm behaviours**
- Integrate `tests/root.yml`  in Travis CI build
- Use latest versions of Ruby, Ansible and docker-compose in Travis CI build environment

pkuczynski commented:
Looks good, but is it worth dropping old systems, instead of keeping them and adding newer ones?

gildegoma commented:
> is it worth dropping old systems [...] ?

Sorry, I forgot to give my rationale about this proposition of removals. 

It is mainly motivated by the fact that these versions are (or will "soon" be) _end of life_ (see summary and references below), and thus `rvm1-ansible` project may not want to support them any longer.
A wild guess, I thought that the goal was here to test and support the two last major releases of each operating system platform.

**EOL Dates Summary:**
- CentOS 6: Full Updates: 20017-05-10 and Maintenance Updates: 2020-11-30
- Debian 8: Full Updates: 2018-06-17 and Security Updates: 2020-06-30
- Ubuntu 16.04: Standard Support: April 2021 and EOL: April 2024
- Ubuntu 14.04: Standard Support: April 2019 and EOL: April 2022

---------------
**References:**

- https://wiki.centos.org/About/Product
- https://wiki.debian.org/LTS
- https://wiki.ubuntu.com/Releases

gildegoma commented:
@pkuczynski I added f3f0952ade49c1a21aa13e3c584a403da1b94f22 as an example, but I'll adapt it as soon as you decide which versions you want to keep in the test matrix.

derekgottlieb commented:
Is it overly aggressive to drop OS versions that are still supported upstream? Should we wait until they become EOL and put up quick PRs to remove them at that time?

gildegoma commented:
As I said, your list will be my list 😉 

PS: I am baking a next PR adding support for Arch Linux 😇.

pkuczynski commented:
I would keep testing old systems even beyond the EOL as long as it's not too costly. There has been many cases in RVM where people took ownership of some very old project and they had to maintain it before they could upgrade it.

In this case, having the older systems does not cost much, so I would keep them in.

gildegoma commented:
😒 second time that the Travis build fails due to "transient problems". See [here an example of rate limiting](https://travis-ci.org/github/rvm/rvm1-ansible/builds/711825750#L24249-L24250) issue on https://github.com.

I'll give a try to parallelise a little bit more. Not only via Ansible from the same run, but with the different travis jobs to try to avoid such problems, and speed up the build time.

gildegoma commented:
I think that a1bd852 is a good move. This will 
- Speed up the CI process (especially if the project tests against many docker images)
- Help to quickly identify the failure causes (which platform or playbook is failing)
- Allow to rebuild only a portion of the total build (very useful at the moment, because of these transient problems, e.g. GPG key installation failure)

![image](https://user-images.githubusercontent.com/434654/88471708-853ac580-cf0c-11ea-8026-e46bc5c8ac16.png)

gildegoma commented:
@pkuczynski I think this PR is now ready for your review.

pkuczynski commented on tests/docker-compose.yml:
```diff
@@ -1,26 +1,35 @@
-    cap_add:
-      - ALL
+
+  centos8:
```
I would prefer to see 8 after 6 not before ;)

gildegoma replied:
Marking as resolved (pending on 6df0663b4097bb79cdced899c52e578f74ec0c00 approval or rejection)

pkuczynski commented on tests/docker-compose.yml:
```diff
@@ -1,26 +1,35 @@
+    build: ./dockerfiles/centos6
+    cap_add: [ALL]
+
+  debian10:
```
Same here, keep ASC version ordering

gildegoma replied:
up to you, but please my comment below, and "double confirm" your decision: ASC (or DESC).

Thank you

gildegoma replied:
Done in 6df0663b4097bb79cdced899c52e578f74ec0c00. Please confirm it 😉

gildegoma replied:
@pkuczynski we keep the Ascending order, right?

pkuczynski commented on tests/inventory:
```diff
@@ -1,9 +1,15 @@
 [centos]
-tests_centos6_1 ansible_connection=docker
+tests_centos8_1 ansible_connection=docker
```
ASC version ordering

gildegoma replied:
Marking as resolved (pending on 6df0663b4097bb79cdced899c52e578f74ec0c00 approval or rejection)

pkuczynski commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@
 
 * Fixes bare variables in conditionals (deprecation warnings with Ansible 2.8+) (#204)
 * Fix boolean values when checking array (#207)
+* Update and enhance the test environment (#215)
```
I believe it would be good to be more expressive what this change brings (nature of changes in the tests, extended OS coverage, etc)

gildegoma replied:
Since there is a reference to the issue/pull-request (#215, commit comments are detailed), I thought that it would be enough (and handy) to keep it short (ideally fitting in a single line). I also had a look at the previous entries in CHANGELOG.md to make my proposal.

What about cherry-picking from these bits 😅:

> - Update the test environment (#205)
>     - Add following target systems: CentOS 8, Debian 10, Ubuntu 18.04 and 20.04
>     - Install Ruby 2.6.6 and 2.7.1 with RVM
> - Modify some test assertions to better abstract RVM behaviour from its internal implementation ("blackbox" testing) (#205)
> - Fix and enhance Travis CI setup (#205)
>      - Install more recent versions of Ruby, to be compatible with latest Bundler version
>      - Validate system-wide setup (root.yml) 
>      - Parallelize the tests to speed up the builds

To avoid a to long ping-pong, could you please give me the text you would put instead? You better know how you want this should look like... 😊

pkuczynski replied:
Looks good! Updated the CHANGELOG

pkuczynski commented:
Possibly we could consider moving to Github Actions in case Travis is not being fast or stable...

gildegoma commented:
@pkuczynski about ASC vs DESC ordering in `docker-compose.yml` and `inventory` files:

I **intentionally** reversed it, with the intention to put the most recent stuff on the top (as you typically do with release notes in a CHANGELOG file). At this occasion, the diff looks a little bit weird, but it won't be the case in the future (if we keep the convention). If you prefer the ASC sorting convention, no problem, but I just needed to give you these details.

gildegoma commented:
> Possibly we could consider moving to Github Actions in case Travis is not being fast or stable...

In this particular case, I don't think that Travis is performing bad at all (I faced similar problems on my local testing, but didn't want to tackle this problem in the scope of this PR). 

That said, moving to GitHub Actions could be an interesting move, but certainly in a distinct pull request 😉

pkuczynski commented:
@thbar can you have a look at this PR please?

thbar commented on .travis.yml:
```diff
@@ -4,37 +4,46 @@ language: generic
-  - ansible-playbook user.yml -l ubuntu
+  - docker-compose config --services | grep $PLATFORM | xargs docker-compose up -d
+  - ansible-playbook $PLAYBOOK -l $PLATFORM
+  - docker-compose down
```
I wonder what will happen if the previous command fails. Will the VM still shut-down? I presume yes, it is more curiosity as I didn't test that type of scenario on Travis.

gildegoma replied:
AFAIK, Travis CI doesn't behave in a "fail-fast" fashion by default, i.e. every statement will be executed (which might not be the desired behaviour, and I'll propose a change on that regard).

See https://github.com/travis-ci/travis-ci/issues/1066

PS: In this very specific example, the `docker-compose down` could (and maybe should) simply be removed as it adds nothing, and slows the build for nothing (it was initially there, when the build "matrix" was still tested within a single sequential job).

gildegoma replied:
@thbar I propose to try "fail fast" (see 943ff233833892bf97be6b149fce09cdd9007f67).

Note: not 100% sure, this will be enough to be applied on `script:` stages as well... (to be verified "soon" ;-)

thbar commented on .travis.yml:
```diff
@@ -4,37 +4,46 @@ language: generic
-    | xargs -n 2 -t sh -c 'test -e $HOME/docker/$1.tar.gz || docker save $0 | gzip -2 > $HOME/docker/$1.tar.gz'
+    mkdir -p $DOCKER_CACHE_DIR && docker images -a --filter='dangling=false' --format '{{.Repository}}:{{.Tag}} {{.ID}}' "tests_$PLATFORM*"
+    | xargs -n 2 -t sh -c 'test -e $DOCKER_CACHE_DIR/$1.tar.gz || docker save $0 | gzip -2 > $DOCKER_CACHE_DIR/$1.tar.gz'
+  - du -hs $DOCKER_CACHE_DIR/*
```
While I can decipher what this does, I think having a little comment to give a clue to other maintainers could be helpful (although comments can become outdated & more dangerous than none :smile:)

gildegoma replied:
hehe... I didn't enciphered much as what this one-liner already was, but that's right that a comment will be welcome. I'll try to put something...

gildegoma replied:
@thbar comments added. Please let me know what do you think...

thbar replied:
@gildegoma looks good enough to me! Thanks for everyone!

thbar commented:
@pkuczynski approved overall, yet the build is failing, so we will need presumably to work on this before merging!

pkuczynski commented:
> > Possibly we could consider moving to Github Actions in case Travis is not being fast or stable...
> 
> In this particular case, I don't think that Travis is performing bad at all (I faced similar problems on my local testing, but didn't want to tackle this problem in the scope of this PR).
> 
> That said, moving to GitHub Actions could be an interesting move, but certainly in a distinct pull request 😉

I did move some of my projects from Travis to GHA and it was pretty smooth yet seemed to be much faster than Travis. Should be done in a separate PR of course...

pkuczynski commented:
> @pkuczynski approved overall, yet the build is failing, so we will need presumably to work on this before merging!

Yeah, I agree. If this supposed to update the test env, it should also make sure that it's green. @gildegoma can you try to fix the failing build?

gildegoma commented on CHANGELOG.md:
```diff
@@ -2,15 +2,31 @@
+* Fix and enhance Travis CI setup (#205)
+  * Install more recent versions of Ruby, to be compatible with latest Bundler version
+  * Validate system-wide setup (root.yml)
+  * Parallelize the tests to speed up the builds
```
@pkuczynski awesome, thank you very much. I like the new "Categorization" grouping.

gildegoma commented:
> The only suggestion I would make (but that will be a good topic for a separate PR) is to have variables for the 2 versions of ruby installed (currently ruby-2.6.6 and ruby-2.7.1), so that we can update in DRY fashion.

I have prepared a patch for this, and will submit it as a new PR, as soon as we get this one merged.

gildegoma commented:
> yet the build is failing, so we will need presumably to work on this before merging!

About the Travis CI failures... These are transient problems (that partially motivated me to split the jobs, so one can restart the failing parts without having to wait for a full rebuild. We'll come back onto this when we'll study a migration to GitHub Actions,...

For the records, I put here some comments about the two jobs that failed ([#395.1](https://travis-ci.org/github/rvm/rvm1-ansible/jobs/713244603) and [#395.2](https://travis-ci.org/github/rvm/rvm1-ansible/jobs/713244604)).

395.1: The [rvm-installer](https://raw.githubusercontent.com/rvm/rvm/master/binscripts/rvm-installer) script couldn't be download (from raw.githubusercontent.com, which returned a 502 error).

I'll propose an _[until-retries-delay](https://www.ansible.com/blog/ansible-tips-and-tricks-dealing-with-unreliable-connections-and-services)_ construct for the related **get_url** task (but again outside of the scope of this PR). Same fix could be applied elsewhere 

![image](https://user-images.githubusercontent.com/434654/89080201-8181c580-d388-11ea-9c5b-491e502e37c8.png)

395.2: This one is quite strange (never seen in previous builds). It looks like docker-compose detects a syntax error in the `docker-compose.yml` file, which is not the case at all.

![image](https://user-images.githubusercontent.com/434654/89080471-3916d780-d389-11ea-8e5e-e55e91f9fbe6.png)

gildegoma commented:
Now, I will kick these two failed jobs, for the example...

gildegoma commented:
⏫ rebased to amend my previous commits (I had GPG signing disabled on my current laptop).

gildegoma commented:
@Kulgar @thbar @pkuczynski I think I have addressed the requested changes, and added a few more improvements. 
I know, that it's getting too large to be easily reviewed, and that's why I would like to get it merged asap. It will unlock other PRs (having CI working), and allow to provide additional improvements via smaller PR iterations.

thbar commented:
I see that @pkuczynski already reviewed, so that's 3 pair of eyes, I'll go ahead and merge.
